packages/controllers/src/utils/UrlUtils.ts

```diff
@@ -0,0 +1,187 @@
+    let end = currentOrigin.indexOf('/', start)
+    if (end === -1) {
+      end = currentOrigin.length
+    }
+    const rawHostPort = currentOrigin.slice(start, end)
+    if (port !== undefined) {
+      return `${host}:${port}` === rawHostPort
+    }
+
+    const rawHostOnly = rawHostPort.split(':')[0]
+    return host === rawHostOnly
+  }
+
+  // Fallback to URL parsing when origin isn't a standard scheme URL
+  const current = parseUrl(currentOrigin)
+  if (!current) {
+    return false
+  }
+  if (port !== undefined) {
+    return host === current.hostname && port === (current.port || undefined)
+  }
+
+  return host === current.hostname
+}
+
+/**
+ * Checks if the current origin matches a wildcard pattern.
+ * @param current - The current origin as a URL object.
+ * @param currentOrigin - The current origin as a string.
+ * @param pattern - The wildcard pattern string to use.
+ * @returns True if matches the wildcard pattern, otherwise false.
+ */
+export function matchWildcardPattern(
+  current: URL,
+  currentOrigin: string,
+  pattern: string
+): boolean {
+  // Extract scheme if present and strip path
+  let working = pattern
+  let scheme: string | undefined = undefined
+  const schemeIdx = working.indexOf('://')
+  if (schemeIdx !== -1) {
+    scheme = working.slice(0, schemeIdx)
+    working = working.slice(schemeIdx + 3)
+  }
+  const slashIdx = working.indexOf('/')
+  if (slashIdx !== -1) {
+    working = working.slice(0, slashIdx)
+  }
+
+  // Split host and optional port
+  let hostPart = working
+  let portPart: string | undefined = undefined
+  const lastColon = hostPart.lastIndexOf(':')
+  if (lastColon !== -1) {
+    portPart = hostPart.slice(lastColon + 1)
+    hostPart = hostPart.slice(0, lastColon)
+  }
+
+  // Validate wildcard usage (only full-label '*')
+  const patternLabels = hostPart.split('.')
+  for (const label of patternLabels) {
+    if (label.includes('*') && label !== '*') {
+      return false
+    }
+  }
+
+  // Scheme must match when specified
+  const currentScheme = current.protocol.replace(/:$/u, '')
+  if (scheme && scheme !== currentScheme) {
+    return false
+  }
+
+  // Port must match exactly when specified (or '*' allows any)
+  if (portPart !== undefined) {
+    if (portPart !== '*' && portPart !== current.port) {
+      return false
+    }
+  }
+
+  /**
+   * Host must have the same number of labels; '*' matches exactly one label.
+   * Uses raw host from the original origin to preserve case-sensitivity.
+   */
+  const raw = parseOriginRaw(currentOrigin)
+  const hostForCompare = raw ? raw.host : current.hostname
+  const currentLabels = hostForCompare.split('.')
+  if (patternLabels.length !== currentLabels.length) {
+    return false
+  }
+
+  for (let i = patternLabels.length - 1; i >= 0; i -= 1) {
+    const p = patternLabels[i]
+    const c = currentLabels[i]
+    if (p !== '*' && p !== c) {
+      return false
+    }
+  }
+
+  return true
+}
+
+export const UrlUtils = {
```

Comment:
If we are exporting them all functions separately, why we need to export `UrlUtils`? Afaik it cannot do tree-shaking if we export all functions from single constant right?